src/components/mobile/survey-questions.tsx

```diff
@@ -82,6 +91,83 @@ export function MobileSurveyQuestionnaire({
     },
   });
 
+  useEffect(() => {
+    let PreviouslyOffline = false;
+    const handleOnline = async () => {
+      if (PreviouslyOffline) {
+        try {
+          await SaveResponsesToDatabase(responses, session, submitResponse);
+          toast({
+            title: "Back online!",
+            description:
+              "Your (intermediate) responses have been submitted successfully.",
+          });
+        } catch (error) {
+          toast({
+            title: "Failed to resend responses",
+            description:
+              "An error occurred while attempting to resend responses.",
+            variant: "destructive",
+          });
+        }
+        PreviouslyOffline = false;
+      }
+    };
+
+    const handleOffline = () => {
+      PreviouslyOffline = true;
+      console.log("Offline - Failed to save responses. Retrying...");
+      // Display error toast if offline
+      toast({
+        title: "Failed to save responses. Retrying...",
+        description:
+          "Data submission in progress... Your responses will be automatically submitted once you're back online. Feel free to continue filling out the survey.",
+        variant: "informative",
+      });
+    };
+    if (!navigator.onLine) {
+      // Handle offline when component mounts
+      handleOffline();
+    } else {
+      handleOnline().catch(() => {
+        toast({
+          title: "Failed to resend responses",
+          description:
+            "An error occurred while attempting to resend responses.",
+          variant: "destructive",
+        });
+      });
+    }
+
+    // Add event listeners for online and offline events
+    window.addEventListener("online", () => {
+      handleOnline().catch(() => {
+        toast({
+          title: "Failed to resend responses",
+          description:
+            "An error occurred while attempting to resend responses.",
+          variant: "destructive",
+        });
+      });
+    });
+    window.addEventListener("offline", handleOffline);
+
+    return () => {
+      // Remove event listeners when component unmounts
+      window.addEventListener("online", () => {
+        handleOnline().catch(() => {
+          toast({
+            title: "Failed to resend responses",
+            description:
+              "An error occurred while attempting to resend responses.",
+            variant: "destructive",
+          });
+        });
+      });
+      window.removeEventListener("offline", handleOffline);
+    };
+  }, [responses, session, submitResponse]);
```

Comment:
I think this effect runs more often than you realize. `submitResponse` seems to be a new object every time, so this effect runs effectively on every cycle.

Reply:
Created #47